[mlir-tensorrt] Transpose Reshape Elimination pass

matthewfl opened the pull request:
The `TransposeReshapeElimination` pass is designed to subsume the existing Transpose and Reshape Elimination passes.  This pass adds a lot of new rewrite rules which enable pushing the transposes and reshapes around so that they can be combined and then eliminated.  The rules from the existing `TransposeElimination` are copied into the `TransposeReshapeElimination.cpp` file.  The rules from the `ReshapeElimination` pass should be subsumed by the rules added to the `TransposeReshapeElimination`.

The motivation for this pass is that there are some cases where shuffles can get inserted around matrix multiplications and element wise ops which break [various fusions inside of TensorRT](https://docs.nvidia.com/deeplearning/tensorrt/latest/performance/best-practices.html#types-of-fusions).

The process is as follows:
1. "canonicalize" the network into simpler ops
    - `shuffle(x)` -> `reshape(transpose(reshape(x)))`
    - `matrix_multiply(x, y)` -> `einsum("ij,jk->ik", x, y)`
    - `expand_rank(x)` -> `reshape(x)`
    - `collapse_rank(x)` -> `reshape(x)`
 2. Push down `reshape` and `transpose` ops as much as possible.  Merging and eliminating when possible
    - `einsum(transpose(x), ...)` -> `einsum(x, ...)` Merge transpose into einsum
    - `einsum(...)` -> `transpose(einsum(...))` Pull transpose out of einsum (to try to match matrix multiply pattern)
    - `einsum(reshape(x), y, ...)` -> `transpose(reshape(einsum(x, reshape(transpose(y)), ...)))` Push reshape down.  Possibly add reshape and transposes to other inputs as needed.  Conditioned on heuristic checking if "better"
    - EXISTING `unary(transpose(x))` -> `transpose(unary(x))`
    - EXISTING `activation(transpose(x))` -> `transpose(activation(x))`
    - EXSITING `identity_op(transpose(x))` -> `transpose(identity_op(x))`
    - `activation(reshape(x))` -> `reshape(activation(x))`
    - `unary(reshape(x))` -> `reshape(unary(x))`
    - `identity_op(reshape(x))` -> `reshape(identity_op(x))`
    - `reshape(transpose(x))` -> `transpose(reshape(x))` if possible put reshape before transpose
    - `qdq(transpose(x))` -> `transpose(qdq(x))` if the scale is 0-dim
    - `qdq(reshape(x))` -> `reshape(qdq(x))` if the scale is 0-dim
    - EXISTING `reshape(reshape(x))` -> `reshape(x)`
    - EXISTING `transpose(transpose(x))` -> `transpose(x)`
    - EXISTING `reshape(x)` -> `x` if `reshape` is identity
    - EXISTING `transpose(x)` -> `x` if `transpose` is identity
    - `elementwise(reshape(a), b)` -> `reshape(elementwise(a, reshape(b)))` conditioned on heuristic
    - EXISTING `elementwise(transpose(a), b)` -> `transpose(elementwise(a, transpose(b)))`
    - `softmax(transpose(x))` -> `transpose(softmax(x))`
    - `softmax(reshape(x))` -> `reshape(softmax(x))`
3. Push up `reshape` and `transpose` ops as much as possible.  Merging and eliminating when possible 
    - `transpose(einsum(...))` -> `einsum(...)`. Merge transpose into einsum
    - `einsum(...)` -> `einsum(transpose(x), ...)`. Pull transposes out of einsum (to try to match matrix multiply pattern)
    - `reshape(einsum(...))` -> `einsum(reshape(transpose(x)), ...)` Push reshapes up through einsum.  Adding transposes as needed
    - EXISTING `transpose(activation(x))` -> `activation(transpose(x))`
    - EXISTING `transpose(unary(x))` -> `unary(transpose(x))`
    - EXISTING `transpose(identity_op(x))` -> `identity_op(transpose(x))`
    - `reshape(activation(x))` -> `activation(reshape(x))`
    - `reshape(unary(x))` -> `unary(reshape(x))`
    - `reshape(identity_op(x))` -> `identity_op(reshape(x))`
    - EXISTING `reshape(reshape(x))` -> `reshape(x)`
    - EXISTING `transpose(transpose(x))` -> `transpose(x)`
    - EXISTING `reshape(x)` -> `x` if `reshape` is identity
    - EXISTING `transpose(x)` -> `x` if `transpose` is identity
    - `transpose(reshape(x))` -> `reshape(transpose(x))` if possible put transpose before reshape
    - `transpose(qdq(x))` -> `qdq(transpose(x))` if the scale is 0-dim
    - `reshape(qdq(x))` -> `qdq(reshape(x))` if the scale is 0-dim
    - `reshape(elementwise(a, b))` -> `elementwise(reshape(a), reshape(b))`
    - EXISTING `transpose(elementwise(a, b))` -> `elementwise(transpose(a), transpose(b))`
    - `transpose(softmax(x))` -> `softmax(transpose(x))`
    - `reshape(softmax(x))` -> `softmax(reshape(x))`
4. Convert back to matrix multiplication form to assist with TRT's pattern matching
    - `einsum(x, y)` -> `matrix_multiply(x, y)` if einsum matches a matrix multiply pattern
    - `matrix_multiply(transpose(x), y)` -> `matrix_multiply(x, y)` merge transpose if possible
5. Final clean ups, additional merging of transpose/reshapes into leftover einsums
    - `einsum(x, y)` -> `matrix_multiply(x, y)` if einsum matches a matrix multiply pattern
    - `matrix_multiply(transpose(x), y)` -> `matrix_multiply(x, y)` merge transpose if possible
    - `transpose(einsum(...))` -> `einsum(...)`
    - `einsum(tranpose(x), ...)` -> `einsum(...)`
    - `einsum(collapse_rank(x), ...)` -> `einsum(...)`
    - `expand_rank(einsum(...))` -> `einsum(...)`
    
    
    
__NOTE__: The overarching goal of this PR is to improve the pattern matching inside of TensorRT (and therefore the quality of kernel's that TensorRT can generate, and fusion that TensorRT will generate).  I have some empirical evidence that the mlir that is generated seems to be be an improvement, however I am still not 100% sure what is the _best_ way to generate mlir in some of these edge cases when it comes to getting the fastest model out of TensorRT.

matthewfl commented:
@christopherbate @shelkesagar29 This is the big PR to eliminate shuffles from the program.  Ideally, this should include deleting the `ReshapeElimination` and `TransposeElimination` and replacing them with the new pass (including updating their lit tests when necessary to match the new pattern).  Not sure what strategy you would like to take to get this merged upstream.

christopherbate commented:
If the pass can subsume the existing two passes, that's fine. I'd say keep the existing test files in-place and just update the RUN command and FileCheck directives -- this way we can see what changed with respect to tests. A follow on commit can merge the test files if required.

matthewfl commented on mlir-tensorrt/tensorrt/test/Dialect/TensorRT/reshape-elimination.mlir:
```diff
@@ -1,4 +1,4 @@
-// RUN: tensorrt-opt %s -split-input-file  -tensorrt-reshape-elimination | FileCheck %s
+// RUN: tensorrt-opt %s -split-input-file  -tensorrt-transpose-reshape-elimination | FileCheck %s
```
@christopherbate I have updated the lit test for reshape and transpose to use the new `transpose-reshape-elimination` pass

christopherbate commented on mlir-tensorrt/tensorrt/lib/TensorRT/Transforms/TransposeReshapeElimination.cpp:
```diff
@@ -0,0 +1,2568 @@
+    for (size_t i = 0; i < op.getInputs().size(); i++) {
+      auto input = op.getInputs()[i];
+      if (auto transpose = input.getDefiningOp<tensorrt::TransposeOp>()) {
+        auto perm = transpose.getPermutation();
```
nit: `r/auto/AffineMap`
https://llvm.org/docs/CodingStandards.html#use-auto-type-deduction-to-make-code-more-readable

christopherbate commented on mlir-tensorrt/tensorrt/lib/TensorRT/Transforms/TransposeReshapeElimination.cpp:
```diff
@@ -0,0 +1,2568 @@
+    if (!einsum)
+      return failure();
+
+    SmallVector<OpOperand> einsumUsers(einsum->getUsers());
```
`if(!einsum.hasOneUse())`

christopherbate commented on mlir-tensorrt/tensorrt/lib/TensorRT/Transforms/TransposeReshapeElimination.cpp:
```diff
@@ -0,0 +1,2568 @@
+    auto newEinsum = rewriter.create<tensorrt::EinsumOp>(
+        op.getLoc(), op.getType(), einsum.getInputs(), newEinsumEquation);
+    rewriter.replaceOp(op, newEinsum.getResult());
+    rewriter.eraseOp(einsum);
```
Don't erase operations in patterns unless absolutely necessary (e.g. they are side-effecting). Erasing pure ops is done automatically and is error prone.

christopherbate commented on mlir-tensorrt/tensorrt/lib/TensorRT/Transforms/TransposeReshapeElimination.cpp:
```diff
@@ -0,0 +1,2568 @@
+  using OpRewritePattern<tensorrt::EinsumOp>::OpRewritePattern;
+  LogicalResult matchAndRewrite(tensorrt::EinsumOp op,
+                                PatternRewriter &rewriter) const override {
+    for (auto input : op.getInputs()) {
```
It's not clear what is the utility of this pattern, especially since one above absorbs the transpose.

It's not clear at all how you're reasoning about the cost of adding a transpose where it doesn't necessarily need to exist.

christopherbate replied:
NVM I see now below that this is related to raising einsum-to-matmul. I made another longer comment about questions with this approach.

christopherbate reviewed (changes requested):
Regarding Matul-to-Einsum and Einsum-to-Matmul:

We should avoid the question of which one is better entirely. I really want to avoid making the TRT dialect transforms reason about which one is better since it's an impossible question  for encoding into a heuristic. It's not easy or straightforward for us to figure out a general rule regarding where einsum might be better since it depends on the internal details of TensorRT, the TensorRT version, etc, and it is not easily to make robust to future changes.

For that reason, I would recommend to users to make this decision much closer to the frontend. For example, in the stablehlo conversions we explicitly have an option for whether or not to prefer use of `tensorrt.einsum` or `tensorrt.matrix_multiply`.

The biggest issue I can see for Einsum vs. MatMul is for issues regarding pattern recognition of special fusions internal to TRT (e.g. MHA) or some either niche feature or optimization that a user may be expecting to see based off of TRT documentation or some other communication. To my knowledge, there's no special pattern that requires einsum, so I would be hesitant to change all matmul to einsum *in the main TRT dialect compilation pipeline* unless its guarded by an option. There are other users who are using TRT dialect directly at the frontend, and this would be a surprising change for them IMO.

That said, from my personal experience, it's Ok to "prefer einsum" as long as you have a mechanism for side-steping any potential issues related to fusion optimizations that might be ciritical to your workloads. In the case of MHA or other critical fusions for popular models right now, we are adding a `tensorrt.attention` operation, which should obviate the issue with respect to einsum vs. matmul.

matthewfl commented:
Thanks for the through review @christopherbate.

I will start working to address the comments on the PR

> We should avoid the question of which one is better entirely. I really want to avoid making the TRT dialect transforms reason about which one is better since it's an impossible question for encoding into a heuristic.

The original motivation for eliminating shuffles/transpose/reshapes was that it broke pattern matching in TensorRT in the first place.  See this bug for additional context: https://partners.nvidia.com/Bug/ViewBug/5381960
WRT to einsum vs matrix multiply, given the design of the PR is to do `matmul -> einsum`, followed by optimizations on the einsum representation, and then converting back `einsum -> matmul`.  This means that what the op was originally (einsum or matrix multiply) is not longer available at the time of `einsum -> matmul` conversion.

Based on observations that I have made of TensorRT, it seems like there is no issue using einsum to represent matrix multiplies, in that it generates the same kernels (but I can't be 100% sure of this given that TRT is closed source and I don't have access to its source).

shelkesagar29 commented:
Pattern `MoveReshapeBeforeTranspose` is causing issue in the following IR.
```
func.func @tensorrt_cluster(%arg0: tensor<1x2x3x3xf32>, %arg1: tensor<1x2x2x2xf32>, %arg2: tensor<1x2x2x3xf16>, %arg3: tensor<3x2x1x1xf16>, %arg4: tensor<2x1x1x1xf32>, %arg5: tensor<1x2x3x3xf32>, %arg6: tensor<1x2x3xf32>, %arg7: tensor<3x2x3xf32>, %arg8: tensor<1x2x3x3x3xf32>, %arg9: tensor<3x2x1x1x1xf32>, %arg10: tensor<1x2x5x5xf32>, %arg11: tensor<1x2x2x2xf32>) -> (tensor<1x1x2x4xf32> {tensorrt.shape_profile = #tensorrt.shape_profile<min = [1, 1, 2, 4], opt = [1, 1, 2, 4], max = [1, 1, 2, 4]>}, tensor<1x3x3x3xf16> {tensorrt.shape_profile = #tensorrt.shape_profile<min = [1, 3, 3, 3], opt = [1, 3, 3, 3], max = [1, 3, 3, 3]>}, tensor<1x2x3x5xf32> {tensorrt.shape_profile = #tensorrt.shape_profile<min = [1, 2, 3, 5], opt = [1, 2, 3, 5], max = [1, 2, 3, 5]>}, tensor<1x1x5x5xf32> {tensorrt.shape_profile = #tensorrt.shape_profile<min = [1, 1, 5, 5], opt = [1, 1, 5, 5], max = [1, 1, 5, 5]>}, tensor<1x1x7x7xf32> {tensorrt.shape_profile = #tensorrt.shape_profile<min = [1, 1, 7, 7], opt = [1, 1, 7, 7], max = [1, 1, 7, 7]>}, tensor<1x1x3x5xf32> {tensorrt.shape_profile = #tensorrt.shape_profile<min = [1, 1, 3, 5], opt = [1, 1, 3, 5], max = [1, 1, 3, 5]>}, tensor<1x3x7xf32> {tensorrt.shape_profile = #tensorrt.shape_profile<min = [1, 3, 7], opt = [1, 3, 7], max = [1, 3, 7]>}, tensor<1x3x3x5x5xf32> {tensorrt.shape_profile = #tensorrt.shape_profile<min = [1, 3, 3, 5, 5], opt = [1, 3, 3, 5, 5], max = [1, 3, 3, 5, 5]>}, tensor<1x1x2x2xf32> {tensorrt.shape_profile = #tensorrt.shape_profile<min = [1, 1, 2, 2], opt = [1, 1, 2, 2], max = [1, 1, 2, 2]>}, tensor<1x1x2x1xf32> {tensorrt.shape_profile = #tensorrt.shape_profile<min = [1, 1, 2, 1], opt = [1, 1, 2, 1], max = [1, 1, 2, 1]>}) attributes {cluster.tensorrt} {
    %0 = tensorrt.slice %arg1[0, 0, 1, 1][1, 2, 2, 2][1, 1, -1, -1] : tensor<1x2x2x2xf32> to tensor<1x2x2x2xf32>
    %1 = tensorrt.transpose {permutation = affine_map<(d0, d1, d2, d3) -> (d1, d0, d2, d3)>} %0 : tensor<1x2x2x2xf32> to tensor<2x1x2x2xf32>
    %2 = tensorrt.deconvolution {dilation = array<i64: 1, 1>, post_padding = array<i64: 1, 1>, pre_padding = array<i64: 1, 1>, stride = array<i64: 1, 2>} in(%arg0 : tensor<1x2x3x3xf32>) kernelWeights(%1 : tensor<2x1x2x2xf32>) -> tensor<1x1x2x4xf32>
    %3 = tensorrt.slice %arg3[0, 0, 0, 0][3, 2, 1, 1][1, 1, -1, -1] : tensor<3x2x1x1xf16> to tensor<3x2x1x1xf16>
    %4 = tensorrt.transpose {permutation = affine_map<(d0, d1, d2, d3) -> (d1, d0, d2, d3)>} %3 : tensor<3x2x1x1xf16> to tensor<2x3x1x1xf16>
    %5 = tensorrt.deconvolution {dilation = array<i64: 1, 1>, post_padding = array<i64: 0, 0>, pre_padding = array<i64: 0, 0>, stride = array<i64: 2, 1>} in(%arg2 : tensor<1x2x2x3xf16>) kernelWeights(%4 : tensor<2x3x1x1xf16>) -> tensor<1x3x3x3xf16>
    %6 = tensorrt.slice %arg4[0, 0, 0, 0][2, 1, 1, 1][1, 1, -1, -1] : tensor<2x1x1x1xf32> to tensor<2x1x1x1xf32>
    %7 = tensorrt.expand_rank %6 : tensor<2x1x1x1xf32> to tensor<2x1x1x1x1xf32>
    %8 = tensorrt.transpose {permutation = affine_map<(d0, d1, d2, d3, d4) -> (d0, d2, d1, d3, d4)>} %7 : tensor<2x1x1x1x1xf32> to tensor<2x1x1x1x1xf32>
    %9 = tensorrt.collapse_rank %8 : tensor<2x1x1x1x1xf32> to tensor<2x1x1x1xf32>
    %10 = tensorrt.deconvolution {dilation = array<i64: 1, 1>, num_groups = 2 : ui32, post_padding = array<i64: 0, 0>, pre_padding = array<i64: 0, 0>, stride = array<i64: 1, 2>} in(%arg0 : tensor<1x2x3x3xf32>) kernelWeights(%9 : tensor<2x1x1x1xf32>) -> tensor<1x2x3x5xf32>
    %11 = tensorrt.slice %arg5[0, 0, 2, 2][1, 2, 3, 3][1, 1, -1, -1] : tensor<1x2x3x3xf32> to tensor<1x2x3x3xf32>
    %12 = tensorrt.transpose {permutation = affine_map<(d0, d1, d2, d3) -> (d1, d0, d2, d3)>} %11 : tensor<1x2x3x3xf32> to tensor<2x1x3x3xf32>
    %13 = tensorrt.deconvolution {dilation = array<i64: 1, 1>, post_padding = array<i64: 1, 1>, pre_padding = array<i64: 1, 1>, stride = array<i64: 2, 2>} in(%arg0 : tensor<1x2x3x3xf32>) kernelWeights(%12 : tensor<2x1x3x3xf32>) -> tensor<1x1x5x5xf32>
    %14 = tensorrt.deconvolution {dilation = array<i64: 1, 1>, post_padding = array<i64: 0, 0>, pre_padding = array<i64: 0, 0>, stride = array<i64: 2, 2>} in(%arg0 : tensor<1x2x3x3xf32>) kernelWeights(%12 : tensor<2x1x3x3xf32>) -> tensor<1x1x7x7xf32>
    %15 = tensorrt.deconvolution {dilation = array<i64: 2, 2>, post_padding = array<i64: 2, 2>, pre_padding = array<i64: 2, 2>, stride = array<i64: 1, 2>} in(%arg0 : tensor<1x2x3x3xf32>) kernelWeights(%12 : tensor<2x1x3x3xf32>) -> tensor<1x1x3x5xf32>
    %16 = tensorrt.expand_rank %arg6 : tensor<1x2x3xf32> to tensor<1x2x1x3xf32>
    %17 = tensorrt.expand_rank %arg7 : tensor<3x2x3xf32> to tensor<3x2x1x3xf32>
    %18 = tensorrt.slice %17[0, 0, 0, 2][3, 2, 1, 3][1, 1, -1, -1] : tensor<3x2x1x3xf32> to tensor<3x2x1x3xf32>
    %19 = tensorrt.transpose {permutation = affine_map<(d0, d1, d2, d3) -> (d1, d0, d2, d3)>} %18 : tensor<3x2x1x3xf32> to tensor<2x3x1x3xf32>
    %20 = tensorrt.deconvolution {dilation = array<i64: 1, 1>, post_padding = array<i64: 0, 0>, pre_padding = array<i64: 0, 0>, stride = array<i64: 1, 2>} in(%16 : tensor<1x2x1x3xf32>) kernelWeights(%19 : tensor<2x3x1x3xf32>) -> tensor<1x3x1x7xf32>
    %21 = tensorrt.collapse_rank %20 : tensor<1x3x1x7xf32> to tensor<1x3x7xf32>
    %22 = tensorrt.slice %arg9[0, 0, 0, 0, 0][3, 2, 1, 1, 1][1, 1, -1, -1, -1] : tensor<3x2x1x1x1xf32> to tensor<3x2x1x1x1xf32>
    %23 = tensorrt.transpose {permutation = affine_map<(d0, d1, d2, d3, d4) -> (d1, d0, d2, d3, d4)>} %22 : tensor<3x2x1x1x1xf32> to tensor<2x3x1x1x1xf32>
    %24 = tensorrt.deconvolution {dilation = array<i64: 1, 1, 1>, post_padding = array<i64: 0, 0, 0>, pre_padding = array<i64: 0, 0, 0>, stride = array<i64: 1, 2, 2>} in(%arg8 : tensor<1x2x3x3x3xf32>) kernelWeights(%23 : tensor<2x3x1x1x1xf32>) -> tensor<1x3x3x5x5xf32>
    %25 = tensorrt.slice %arg10[0, 0, 1, 1][1, 2, 3, 3][1, 1, 1, 1] : tensor<1x2x5x5xf32> to tensor<1x2x3x3xf32>
    %26 = tensorrt.convolution {dilation = array<i64: 1, 1>, post_padding = array<i64: 0, 0>, pre_padding = array<i64: 0, 0>, stride = array<i64: 1, 1>} in(%25 : tensor<1x2x3x3xf32>) kernel(%arg11 : tensor<1x2x2x2xf32>) -> tensor<1x1x2x2xf32>
    %27 = tensorrt.convolution {dilation = array<i64: 1, 2>, post_padding = array<i64: 0, 0>, pre_padding = array<i64: 0, 0>, stride = array<i64: 1, 1>} in(%25 : tensor<1x2x3x3xf32>) kernel(%arg11 : tensor<1x2x2x2xf32>) -> tensor<1x1x2x1xf32>
    return %2, %5, %10, %13, %14, %15, %21, %24, %26, %27 : tensor<1x1x2x4xf32>, tensor<1x3x3x3xf16>, tensor<1x2x3x5xf32>, tensor<1x1x5x5xf32>, tensor<1x1x7x7xf32>, tensor<1x1x3x5xf32>, tensor<1x3x7xf32>, tensor<1x3x3x5x5xf32>, tensor<1x1x2x2xf32>, tensor<1x1x2x1xf32>
}
```
Before deconv at `%10`, we are changing `reshape tensor<2x1x1x1x1> to tensor<2x1x1x1>` to `reshape tensor<2x1x1x1x1> to tensor<2x1>`.

shelkesagar29 reviewed (changes requested):
Currently there are two issues because of which e2e models like whisper Jax and GPT are failing. 
1. In `MoveReshapeBeforeTranspose` pattern. I have commented there with minimal failing test case.
2. In `EinsumPushDownTranspose` pattern. I have commented with failing test and how to fix the issue.

Once these two are fixed, we are good to merge this one.
This is really good work. Thanks

matthewfl commented on mlir-tensorrt/tensorrt/lib/TensorRT/Transforms/TransposeReshapeElimination.cpp:
```diff
@@ -0,0 +1,2480 @@
+  int64_t cost2 = memoryCost(consumer.getType()) + memoryCost(op1.getType());
+  LLVM_DEBUG(DBGS() << "cost1=" << cost1 << ", cost2=" << cost2 << "\n");
+  if (cost1 == 0 && cost2 == 0)
+    return {};
```
This was causing ping-ponging and a non-terminating infinite loop on the following test
https://github.com/NVIDIA/TensorRT-Incubator/pull/686/commits/c3b4d85fb64466e11c825e6b7419e2dda16c150c#diff-e275a939421ea167dbf564d5fe9b866e458b4ee87f70c39e08cccbd333d7b44eR474

matthewfl commented on mlir-tensorrt/tensorrt/lib/TensorRT/Transforms/TransposeReshapeElimination.cpp:
```diff
@@ -0,0 +1,2480 @@
+    Value newTransposeOp = rewriter.createOrFold<tensorrt::TransposeOp>(
+        op.getLoc(), newReshapeOp,
+        AffineMap::getPermutationMap(newTranspose, op.getContext()));
+
```
thanks I will look into these failures

matthewfl commented on mlir-tensorrt/tensorrt/lib/TensorRT/Transforms/TransposeReshapeElimination.cpp:
```diff
@@ -0,0 +1,2480 @@
+    for (auto &[c, i] : outputAxes) {
+      newEinsumRhs += c;
+      newEinsumShape.push_back(op.getType().getDimSize(i));
+      outputPerm.push_back(i);
```
Thanks, I fixed this also

matthewfl commented:
@shelkesagar29 Hi Sagar, I pushed the code that makes sure that the multihead attention fusion is able to still work.  You should be able to test it on your models now.

Let me know if you want me to rebase this on the main branch or squash the commits.  I only added new commits so it should be easier to merge with the copy you were testing.

matthewfl commented:
@shelkesagar29 Thanks for the report.  I just pushed a fixed for the issue that you reported.

matthewfl commented:
@shelkesagar29 I just pushed a fix for the bug you sent via email.

shelkesagar29 commented on mlir-tensorrt/tensorrt/lib/TensorRT/Transforms/TransposeReshapeElimination.cpp:
```diff
@@ -0,0 +1,3135 @@
+
+  std::string generateEquation() const {
+    std::string ret = lhsParts[0];
+    for (size_t i = 1; i < lhsParts.size(); i++) {
```
Remove braces `{}`

shelkesagar29 commented on mlir-tensorrt/tensorrt/lib/TensorRT/Transforms/TransposeReshapeElimination.cpp:
```diff
@@ -0,0 +1,3135 @@
+      return failure();
+
+    SmallVector<std::pair<char, int64_t>> outputAxes;
+    for (size_t i = 0; i < equation.rhs.size(); i++) {
```
Remove braces

shelkesagar29 commented on mlir-tensorrt/tensorrt/lib/TensorRT/Transforms/TransposeReshapeElimination.cpp:
```diff
@@ -0,0 +1,3135 @@
+    LLVM_DEBUG({
+      std::stringstream out;
+      out << "outputAxes: [";
+      for (auto x : outputAxes) {
```
Remove braces

matthewfl commented:
@shelkesagar29 I updated the PR. Cleaned up the formatting.  Squashed the commits down to a single commit and rebased on the current `main` branch.  The commit is also signed with my gpg key.

You can go ahead and click merge on the PR.

shelkesagar29 commented:
> @shelkesagar29 I updated the PR. Cleaned up the formatting. Squashed the commits down to a single commit and rebased on the current `main` branch. The commit is also signed with my gpg key.
> 
> You can go ahead and click merge on the PR.

Done. Thanks for all the hard work and patience
